dspy/functional/typed_predictor_signature.py

```diff
@@ -0,0 +1,130 @@
+            prefix_instructions += "\n\n"
+        instructions = prefix_instructions + "Use only the available information to extract the output fields.\n\n"
+        dspy_fields = {}
+        for field_name, field in pydantic_class_for_dspy_input_fields.model_fields.items():
+            if field.default and 'typing.Annotated' in str(field.default):
+                raise ValueError(f"Field '{field_name}' is annotated incorrectly. See 'Constraints on compound types' in https://docs.pydantic.dev/latest/concepts/fields/")
+
+            is_default_value_specified, is_marked_as_optional, inner_field = cls._process_field(field)
+            if is_marked_as_optional:
+                if field.default is None or field.default is PydanticUndefined:
+                    field.default = 'null'
+                field.description = inner_field.description
+                field.examples = inner_field.examples
+                field.metadata = inner_field.metadata
+                field.json_schema_extra = inner_field.json_schema_extra
+            else:
+                field.validate_default = False
+
+            input_field = InputField(desc=field.description)
+            dspy_fields[field_name] = (field.annotation, input_field)
+
+        for field_name, field in pydantic_class_for_dspy_output_fields.model_fields.items():
+            if field.default and 'typing.Annotated' in str(field.default):
+                raise ValueError(f"Field '{field_name}' is annotated incorrectly. See 'Constraints on compound types' in https://docs.pydantic.dev/latest/concepts/fields/")
+
+            is_default_value_specified, is_marked_as_optional, inner_field = cls._process_field(field)
+            if is_marked_as_optional:
+                if field.default is None or field.default is PydanticUndefined:
+                    field.default = 'null'
+                field.description = inner_field.description
+                field.examples = inner_field.examples
+                field.metadata = inner_field.metadata
+                field.json_schema_extra = inner_field.json_schema_extra
+            else:   
+                field.validate_default = False
+
+            if field.default is PydanticUndefined:
+                raise ValueError(
+                    f"Field '{field_name}' has no default value. Required fields must have a default value. "
+                    "Change the field to be Optional or specify a default value."
+                )
+            
+            output_field = OutputField(desc=field.description if field.description else "")
+            dspy_fields[field_name] = (field.annotation, output_field)
+
+            instructions += f"When extracting '{field_name}':\n"
+            instructions += f"If it is not mentioned in the input fields, return: '{field.default}'. "
+
+            examples = field.examples
+            if examples:
+                quoted_examples = [f"'{example}'" for example in examples]
```

Comment:
quotes will fail on complex values?

Reply:
Only issue I have seen is the LM sometimes returning string values enclosed in quotes. For example, returning "'John Doe'", instead of "John Doe". But this is easy to strip. It correctly extracts numbers without putting them in quotes but I guess the prompt could be enhanced to detect if the example values should be quoted. Any other suggestions on how this could be improved?

Also,  re. complex types - isn't that what I am testing with complex Input and Output BaseModels? So may be the code for pulling apart complex types and building a custom signature is unavoidable. Thoughts?